spring-cloud-stream-binders/spring-cloud-stream-binder-gemfire/src/main/java/org/springframework/cloud/stream/binder/gemfire/MessageKey.java

```diff
@@ -0,0 +1,229 @@
+		this.routingHash = this.hashCode();
+	}
+
+	/**
+	 * Construct a {@code MessageKey} with a specific hash for selecting the
+	 * partitioned region bucket.
+	 *
+	 * @param sequenceId unique sequence id used for ordering messages
+	 * @param producerTimestamp time stamp of when the producer process started
+	 * @param producerPid PID of producer process
+	 * @param routingHash routing hash for bucket selection
+	 */
+	public MessageKey(long sequenceId, long producerTimestamp, int producerPid, int routingHash) {
+		this.sequenceId = sequenceId;
+		this.producerId = producerId(producerTimestamp, producerPid);
+		this.routingHash = routingHash;
+	}
+
+	/**
+	 * Return a producer id consisting of the timestamp of the process startup
+	 * time for the producer and the PID of the producer.
+	 * <p>
+	 * The first (leftmost) 16 bits are reserved for the PID of the producer
+	 * process, rightmost 48 bits are reserved for the timestamp of when the
+	 * producer process was started.
+	 *
+	 * @param producerTimestamp startup timestamp for producer
+	 * @param producerPid PID for producer
+	 * @return producer id
+	 */
+	private long producerId(long producerTimestamp, int producerPid) {
+		long longPid = producerPid;
+		longPid = longPid << 48;
+		return longPid ^ producerTimestamp;
+	}
+
+	/**
+	 * Return the sequence id used for ordering messages.
+	 *
+	 * @return sequence id
+	 */
+	public long getSequenceId() {
+		return sequenceId;
+	}
+
+	/**
+	 * Return the producer id for this message key.
+	 *
+	 * @return producer id
+	 */
+	public long getProducerId() {
+		return producerId;
+	}
+
+	/**
+	 * Return the PID of the producer for this message key.
+	 *
+	 * @return producer PID
+	 */
+	public int getProducerPid() {
+		long longPid = producerId >>> 48;
+		return (int) longPid;
+	}
+
+	/**
+	 * Return the timestamp of when the producer of this message key started.
+	 *
+	 * @return producer startup timestamp
+	 */
+	public long getProducerTimestamp() {
+		return producerId << 16 >> 16;
+	}
+
+	@Override
+	public int compareTo(MessageKey that) {
+		int c = Long.compare(this.getProducerTimestamp(), that.getProducerTimestamp());
+		return (c == 0 ? Long.compare(this.sequenceId, that.sequenceId) : c);
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) return true;
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+
+		MessageKey that = (MessageKey) o;
+		return this.sequenceId == that.sequenceId
+				&& this.producerId == that.producerId;
+	}
+
+	@Override
+	public int hashCode() {
+		int result = (int) (sequenceId ^ (sequenceId >>> 32));
+		result = 31 * result + (int) (producerId ^ (producerId >>> 32));
+		result = 31 * result + routingHash;
+		return result;
+	}
+
+	@Override
+	public String toString() {
+		return "MessageKey{" +
+				"id=" + sequenceId +
+				", producerId=" + producerId +
+				'}';
+	}
+
+	@Override
+	public Object getRoutingObject(EntryOperation<MessageKey, Message<?>> entryOperation) {
```

Comment:
That's an interesting idea. I'm not sure how often it will be invoked either. My thought is to consume as little memory as possible, but repeated boxing would also be inefficient. My guess is that it won't be called very often; it would just be used to determine the hash for selecting a bucket. I'll look into it.